examples/ifm/ifm.py

```diff
@@ -0,0 +1,354 @@
+            r2_score = np.mean(eval_metric.compute_metric("r2"))
+            return {"rmse": rmse_score, "mae": mae_score, "r2": r2_score}
+        else:
+            roc_score = np.mean(eval_metric.compute_metric(metric))
+            prc_score = np.mean(eval_metric.compute_metric("prc_auc"))
+            return {"roc_auc": roc_score, "prc_auc": prc_score}
+
+    return val_loss_func
+
+
+def train(cfg: DictConfig):
+    # set random seed for reproducibility
+    ppsci.utils.misc.set_random_seed(cfg.seed)
+    # initialize logger
+    logger.init_logger("ppsci", os.path.join(cfg.output_dir, "train.log"), "info")
+
+    if cfg.data_label in ["esol", "freesolv", "lipop"]:
+        # task_type = "reg"
+        reg = True
+        metric = "rmse"
+    else:
+        # task_type = "cla"
+        reg = False
+        metric = "roc_auc"
+
+    # set dataloader config
+    train_dataloader_cfg = {
+        "dataset": {
+            "name": "IFMMoeDataset",
+            "input_keys": ("x",),
+            "label_keys": (
+                "y",
+                "mask",
+            ),
+            "data_dir": cfg.data_dir,
+            "data_mode": "train",
+            "data_label": cfg.data_label,
+        },
+        "batch_size": cfg.TRAIN.batch_size,
+        "sampler": {
+            "name": "BatchSampler",
+            "drop_last": False,
+            "shuffle": True,
+        },
+        "num_workers": 1,
+    }
+
+    # set constraint
+    sup_constraint = ppsci.constraint.SupervisedConstraint(
+        train_dataloader_cfg,
+        output_expr={"pred": lambda out: out["pred"]},
+        loss=ppsci.loss.FunctionalLoss(get_train_loss_func(reg)),
+        name="Sup",
+    )
+
+    # parmas from dataset
+    inputs = sup_constraint.data_loader.dataset.data_tr_x.shape[1]
+    tasks = sup_constraint.data_loader.dataset.task_dict[cfg.data_label]
+    iters_per_epoch = len(sup_constraint.data_loader)
+    print(f"inputs is: {inputs}, iters_per_epoch: {iters_per_epoch}")
+    if not reg:
+        pos_weights = sup_constraint.data_loader.dataset.pos_weights
+        sup_constraint.loss = ppsci.loss.FunctionalLoss(
+            get_train_loss_func(reg, pos_weights)
+        )
+
+    # wrap constraints together
+    constraint = {sup_constraint.name: sup_constraint}
+
+    hyper_paras = cfg.HYPER_OPT[cfg.data_label]
+
+    hidden_units = [
+        hyper_paras["hidden_unit1"],
+        hyper_paras["hidden_unit2"],
+        hyper_paras["hidden_unit3"],
+    ]
+    # set model
+    model = ppsci.arch.IFMMLP(
+        # **cfg.MODEL,
+        input_keys=("x",),
+        output_keys=("pred",),
+        hidden_units=hidden_units,
+        embed_name=cfg.MODEL.embed_name,
+        inputs=inputs,
+        outputs=len(tasks),
+        d_out=hyper_paras["d_out"],
+        sigma=hyper_paras["sigma"],
+        dp_ratio=hyper_paras["dropout"],
+        reg=reg,
+        first_omega_0=hyper_paras["omega0"],
+        hidden_omega_0=hyper_paras["omega1"],
+    )
```

Comment:
这里建议通过yaml里的引用语法简化一下
```suggestion
    hyper_paras = cfg.HYPER_OPT[cfg.data_label]

    hidden_units = [
        hyper_paras["hidden_unit1"],
        hyper_paras["hidden_unit2"],
        hyper_paras["hidden_unit3"],
    ]
    # set model
    model = ppsci.arch.IFMMLP(
        # **cfg.MODEL,
        input_keys=("x",),
        output_keys=("pred",),
        hidden_units=hidden_units,
        embed_name=cfg.MODEL.embed_name,
        inputs=inputs,
        outputs=len(tasks),
        d_out=hyper_paras["d_out"],
        sigma=hyper_paras["sigma"],
        dp_ratio=hyper_paras["dropout"],
        reg=reg,
        first_omega_0=hyper_paras["omega0"],
        hidden_omega_0=hyper_paras["omega1"],
    )
```